Missing NexStar Commands:  "J", "h", "P" and "m"

There are several unsupported commands that are continuously spamming the CLI.  These commands are emitted by Stellarium Mobile.  It does not seem like Stellarium Mobile is able to sync the telescope or set the location without knowing the alignment state.  Furthermore, Stellarium Mobile will occasionally query the current telescope time.

The two unimplemented commands are:

### "h" - Get Time
This should return the time in the same format as the "H" command handled at https://github.com/synfinatic/alpacascope/blob/main/telescope/nexstar.go#L219

### "J" - Is alignment complete?
Should return `1` if the alignment is complete.  For the sake of compatibility, this should return `1` if the alignment state is unknown.

Details of these commands may be found in the NexStar protocol document found here:  https://s3.amazonaws.com/celestron-site-support-files/support_files/1154108406_nexstarcommprot.pdf

I did attempt to use the LX200 protocol, but there is an entirely separate issue that crops up:

```
INFO[0000] Looking for Alpaca Remote Server locally on port 11111...
INFO[0003] Found Alpaca on 192.168.30.17:11111
INFO[0003] Starting SkyFi Discovery service on UDP/4031
INFO[0003] Connected to telescope 0:
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/synfinatic/alpacascope/alpaca.(*Telescope).GetAxisRates(0xc0000920c0, 0x0?)
        /Users/aturner/go/src/github.com/synfinatic/alpacascope/alpaca/telescope.go:179 +0x4e5
main.main()
        /Users/aturner/go/src/github.com/synfinatic/alpacascope/cmd/main.go:187 +0xfe5
```

If necessary, I can open a new issue about this as well.

--- a/.github/workflows/go.yaml
+++ b/.github/workflows/go.yaml
@@ -10,17 +10,17 @@ jobs:
     - name: Install X11/OpenGL
       run: sudo apt-get update && sudo apt-get install -y xorg-dev
 
-    - name: Set up Go 1.15
+    - name: Set up Go 1.18
       uses: actions/setup-go@v1
       with:
-        go-version: 1.15
+        go-version: 1.18
       id: go
 
-    - name: Install Fyne tools
-      run: go get fyne.io/fyne/v2/cmd/fyne && go get github.com/fyne-io/fyne-cross
-        
     - name: Check out code into the Go module directory
       uses: actions/checkout@v2
+
+    - name: Install Fyne tools
+      run: make install-fyne && make install-fyne-cross
 
     - name: Tests
       run: make test

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,7 +18,8 @@ Fixed:
 
  - Fix decoding of ISO8601 timestamps which include fractional seconds
     for GetUTCDate
- - Incorrectly returned Lat twice instead of lat/long
+ - Incorrectly returned Lat twice instead of lat/long 
+ - Error out if unable to connect to telescope
 
 ## v2.3.0 - 2022-07-21
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,11 @@ Changed:
 
  - Require Go v1.18+
 
+Fixed:
+
+ - Fix decoding of ISO8601 timestamps which include fractional seconds
+    for GetUTCDate
+
 ## v2.3.0 - 2022-07-21
 
 Added:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@ Added:
 
  - Added missing support for NexStar 'h', 'J' and 'm' commands
  - Added NexStar 'P' command support for GPS, RTC and get device version (1.6)
+ - Improved `--debug` output
 
 Changed:
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,19 +1,23 @@
 # AlpacaScope Changelog
 
-## v2.4.0 - 2022-09-06
+## v2.4.0 - 2022-09-20
 
 Added:
 
- - Support for NexStar 'h' and 'J' commands
+ - Added missing support for NexStar 'h', 'J' and 'm' commands
+ - Added NexStar 'P' command support for GPS, RTC and get device version (1.6)
 
 Changed:
 
  - Require Go v1.18+
+ - Many bug fixes in NexStar code for Stellarium Mobile support
+ - NexStar tracking mode now reports correctly
 
 Fixed:
 
  - Fix decoding of ISO8601 timestamps which include fractional seconds
     for GetUTCDate
+ - Incorrectly returned Lat twice instead of lat/long
 
 ## v2.3.0 - 2022-07-21
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,15 @@
 # AlpacaScope Changelog
 
+## v2.4.0 - 2022-09-06
+
+Added:
+
+ - Support for NexStar 'h' and 'J' commands
+
+Changed:
+
+ - Require Go v1.18+
+
 ## v2.3.0 - 2022-07-21
 
 Added:

--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,7 @@ endif
 BUILDINFOSDET ?=
 PROGRAM_ARGS ?=
 
-PROJECT_VERSION           := 2.3.0
+PROJECT_VERSION           := 2.4.0
 BUILD_ID                  := 1
 DOCKER_REPO               := synfinatic
 PROJECT_NAME              := alpacascope
@@ -75,13 +75,19 @@ build-gui: darwin-gui darwin-release-gui windows linux-gui ## Build GUI binaries
 	@if test -z "`echo $(GOOS) | grep 'mingw64'`" ; then echo "$(MAKECMDGOALS) requires building on Windows/MINGW64" ; exit 1 ; fi
 
 
+install-fyne:  ## Download and install Fyne
+	go install fyne.io/fyne/v2/cmd/fyne@v2.2.2
+
+install-fyne-cross:  ## Download and install Fyne-Cross
+	go install github.com/fyne-io/fyne-cross@v1.3.0
+
 # Install fyne binary in $GOPATh/bin
 .PHONY: .fyne .fyne-cross
 .fyne:
-	@if test -z "`which fyne`"; then echo "Please install fyne: go get fyne.io/fyne/v2/cmd/fyne" ; exit 1 ; fi
+	@if test -z "`which fyne`"; then echo "Please install fyne: make install-fyne" ; exit 1 ; fi
 
 .fyne-cross:
-	@if test -z "`which fyne-cross`"; then echo "Please install fyne-cross: go get github.com/fyne-io/fyne-cross" ; exit 1 ; fi
+	@if test -z "`which fyne-cross`"; then echo "Please install fyne-cross: make install-fyne-cross" ; exit 1 ; fi
 
 # used by our github action to test building the release binaries + GUI on Linux
 .build-test-binaries: $(LINUX_BIN) $(DARWIN_BIN) $(WINDOWS)
@@ -99,7 +105,7 @@ clean-go: ## Clean Go cache
 	go clean -i -r -cache -modcache
 
 go-get:  ## Get our go modules
-	go get -v all
+	go install -v all
 
 .PHONY: build-race
 build-race: .prepare ## Build race detection binary

--- a/alpaca/alpaca.go
+++ b/alpaca/alpaca.go
@@ -9,9 +9,8 @@ package alpaca
 import (
 	"fmt"
 
-	log "github.com/sirupsen/logrus"
-
 	"github.com/go-resty/resty/v2"
+	log "github.com/sirupsen/logrus"
 )
 
 type Alpaca struct {

--- a/alpaca/alpaca.go
+++ b/alpaca/alpaca.go
@@ -9,6 +9,7 @@ package alpaca
 import (
 	"fmt"
 
+	"github.com/davecgh/go-spew/spew"
 	"github.com/go-resty/resty/v2"
 	log "github.com/sirupsen/logrus"
 )
@@ -69,6 +70,7 @@ func (a *Alpaca) GetString(device string, id uint32, api string) (string, error)
 		a.ErrorMessage = resp.String()
 	}
 	result := (resp.Result().(*stringResponse))
+	log.Debugf("Alpaca response: %s", spew.Sdump(result))
 	return result.Value, nil
 }
 
@@ -94,6 +96,7 @@ func (a *Alpaca) GetStringList(device string, id uint32, api string) ([]string, 
 		a.ErrorMessage = resp.String()
 	}
 	result := (resp.Result().(*stringlistResponse))
+	log.Debugf("Alpaca response: %s", spew.Sdump(result))
 	return result.Value, nil
 }
 
@@ -119,6 +122,7 @@ func (a *Alpaca) GetBool(device string, id uint32, api string) (bool, error) {
 		a.ErrorMessage = resp.String()
 	}
 	result := (resp.Result().(*boolResponse))
+	log.Debugf("Alpaca response: %s", spew.Sdump(result))
 	return result.Value, nil
 }
 
@@ -144,6 +148,7 @@ func (a *Alpaca) GetInt32(device string, id uint32, api string) (int32, error) {
 		a.ErrorMessage = resp.String()
 	}
 	result := (resp.Result().(*int32Response))
+	log.Debugf("Alpaca response: %s", spew.Sdump(result))
 	return result.Value, nil
 }
 
@@ -169,6 +174,7 @@ func (a *Alpaca) GetFloat64(device string, id uint32, api string) (float64, erro
 		a.ErrorMessage = resp.String()
 	}
 	result := (resp.Result().(*float64Response))
+	log.Debugf("Alpaca response: %s", spew.Sdump(result))
 	return result.Value, nil
 }
 
@@ -194,6 +200,7 @@ func (a *Alpaca) GetListUint32(device string, id uint32, api string) ([]uint32, 
 		a.ErrorMessage = resp.String()
 	}
 	result := (resp.Result().(*listUint32Response))
+	log.Debugf("Alpaca response: %s", spew.Sdump(result))
 	return result.Value, nil
 
 }
@@ -235,6 +242,7 @@ type putResponse struct {
 
 func (a *Alpaca) Put(device string, id uint32, api string, form map[string]string) error {
 	url := a.url(device, id, api)
+	log.Debugf("Alpaca PUT: %v", spew.Sdump(form))
 	resp, err := a.client.R().
 		SetHeader("Content-Type", "application/x-www-form-urlencoded").
 		SetResult(&putResponse{}).
@@ -252,6 +260,6 @@ func (a *Alpaca) Put(device string, id uint32, api string, form map[string]strin
 	if result.ErrorNumber != 0 {
 		return fmt.Errorf("%d: %s", result.ErrorNumber, result.ErrorMessage)
 	}
-	log.Debugf("%v", result)
+	log.Debugf("Alpaca PUT response: %s", spew.Sdump(result))
 	return nil
 }

--- a/alpaca/telescope.go
+++ b/alpaca/telescope.go
@@ -7,6 +7,8 @@ package alpaca
 import (
 	"fmt"
 	"time"
+
+	"github.com/relvacode/iso8601"
 )
 
 type TrackingMode int
@@ -142,12 +144,16 @@ func (t *Telescope) GetTracking() (TrackingMode, error) {
 	return t.Tracking, nil
 }
 
+// Parse ISO8601 w/ fractional seconds
 func (t *Telescope) GetUTCDate() (time.Time, error) {
-	iso8601, err := t.alpaca.GetString("telescope", t.Id, "utcdate")
+	isoTime, err := t.alpaca.GetString("telescope", t.Id, "utcdate")
 	if err != nil {
 		return time.Unix(0, 0), err
+	} else if isoTime == "" {
+		// sometimes we get no error, but we get an empty string?
+		return time.Unix(0, 0), fmt.Errorf("got an empty UTCDate string")
 	}
-	return time.Parse("2006-01-02T15:04:05Z0700", iso8601)
+	return iso8601.ParseString(isoTime)
 }
 
 type mapAxisRates struct {

--- a/alpaca/telescope.go
+++ b/alpaca/telescope.go
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	"github.com/relvacode/iso8601"
+	log "github.com/sirupsen/logrus"
 )
 
 type TrackingMode int
@@ -118,7 +119,6 @@ func (t *Telescope) GetSlewing() (bool, error) {
 
 func (t *Telescope) GetSiteLatitude() (float64, error) {
 	return t.alpaca.GetFloat64("telescope", t.Id, "sitelatitude")
-
 }
 
 func (t *Telescope) GetSiteLongitude() (float64, error) {
@@ -176,6 +176,14 @@ func (t *Telescope) GetAxisRates(axis AxisType) (map[string]float64, error) {
 		return map[string]float64{}, err
 	}
 	result := (resp.Result().(*mapAxisRates))
+	if len(result.Value) == 0 {
+		log.Errorf("telescope driver returned an empty list for axisrates")
+		// sometime it's an empty list?
+		return map[string]float64{
+			"Minimum": 0.0,
+			"Maximum": 0.0,
+		}, nil
+	}
 	return result.Value[0], nil
 }
 

--- a/cmd/main.go
+++ b/cmd/main.go
@@ -86,8 +86,12 @@ func main() {
 
 	// turn on debugging?
 	if debug == true {
-		log.SetReportCaller(true)
+		log.SetFormatter(&log.TextFormatter{
+			// DisableColors: true,
+			FullTimestamp: true,
+		})
 		log.SetLevel(log.DebugLevel)
+		// log.SetReportCaller(true)
 	} else {
 		// pretty console output
 		log.SetLevel(log.InfoLevel)

--- a/cmd/main.go
+++ b/cmd/main.go
@@ -170,6 +170,13 @@ func main() {
 		if err != nil {
 			log.Fatalf("Unable to connect to telescope ID %d: %s", telescopeId, err.Error())
 		}
+		connected, err = scope.GetConnected()
+		if err != nil {
+			log.Fatalf("Unable to determine status of telescope: %s", err.Error())
+		}
+		if !connected {
+			log.Fatalf("Telescope is not connected to ASCOM Remote")
+		}
 	}
 
 	name, err := scope.GetName()

--- a/go.mod
+++ b/go.mod
@@ -1,16 +1,45 @@
 module github.com/synfinatic/alpacascope
 
-go 1.16
+go 1.18
 
+// when you update the Fyne version, also update the Makefile target
 require (
 	fyne.io/fyne/v2 v2.2.2
 	github.com/go-resty/resty/v2 v2.3.0
-	github.com/kr/text v0.2.0 // indirect
 	github.com/libp2p/go-reuseport v0.1.0
 	github.com/mattn/go-colorable v0.1.8
 	github.com/sirupsen/logrus v1.8.1
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.7.2
 	golang.org/x/sys v0.0.0-20220412211240-33da011f77ad
+)
+
+require (
+	fyne.io/systray v1.10.1-0.20220621085403-9a2652634e93 // indirect
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/fredbi/uri v0.0.0-20181227131451-3dcfdacbaaf3 // indirect
+	github.com/fsnotify/fsnotify v1.5.4 // indirect
+	github.com/fyne-io/gl-js v0.0.0-20220119005834-d2da28d9ccfe // indirect
+	github.com/fyne-io/glfw-js v0.0.0-20220120001248-ee7290d23504 // indirect
+	github.com/fyne-io/image v0.0.0-20220602074514-4956b0afb3d2 // indirect
+	github.com/go-gl/gl v0.0.0-20211210172815-726fda9656d6 // indirect
+	github.com/go-gl/glfw/v3.3/glfw v0.0.0-20211213063430-748e38ca8aec // indirect
+	github.com/godbus/dbus/v5 v5.1.0 // indirect
+	github.com/goki/freetype v0.0.0-20181231101311-fa8a33aabaff // indirect
+	github.com/gopherjs/gopherjs v1.17.2 // indirect
+	github.com/jsummers/gobmp v0.0.0-20151104160322-e2ba15ffa76e // indirect
+	github.com/kr/text v0.2.0 // indirect
+	github.com/mattn/go-isatty v0.0.12 // indirect
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/srwiley/oksvg v0.0.0-20200311192757-870daf9aa564 // indirect
+	github.com/srwiley/rasterx v0.0.0-20200120212402-85cb7272f5e9 // indirect
+	github.com/tevino/abool v1.2.0 // indirect
+	github.com/yuin/goldmark v1.4.0 // indirect
+	golang.org/x/image v0.0.0-20220601225756-64ec528b34cd // indirect
+	golang.org/x/mobile v0.0.0-20211207041440-4e6c2922fdee // indirect
+	golang.org/x/net v0.0.0-20210805182204-aaa1db679c0d // indirect
+	golang.org/x/text v0.3.7 // indirect
 	gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+	honnef.co/go/js/dom v0.0.0-20210725211120-f030747120f2 // indirect
 )

--- a/go.mod
+++ b/go.mod
@@ -8,6 +8,7 @@ require (
 	github.com/go-resty/resty/v2 v2.3.0
 	github.com/libp2p/go-reuseport v0.1.0
 	github.com/mattn/go-colorable v0.1.8
+	github.com/relvacode/iso8601 v1.1.0
 	github.com/sirupsen/logrus v1.8.1
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.7.2

--- a/gui/main.go
+++ b/gui/main.go
@@ -281,8 +281,17 @@ func (c *AlpacaScopeConfig) Run() {
 	}
 
 	if !connected {
+		// Manually connect
 		err = scope.PutConnected(true)
 		if err != nil {
+			sbox.AddLine(fmt.Sprintf("Unable to connect to telescope ID %s", c.AscomTelescope))
+			sbox.AddLine(err.Error())
+			sbox.AddLine(CHECK)
+			tempQuit <- true
+			return
+		}
+		connected, err = scope.GetConnected()
+		if err != nil || !connected {
 			sbox.AddLine(fmt.Sprintf("Unable to connect to telescope ID %s", c.AscomTelescope))
 			sbox.AddLine(err.Error())
 			sbox.AddLine(CHECK)

--- a/telescope/nexstar.go
+++ b/telescope/nexstar.go
@@ -220,22 +220,22 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		// get date/time
 		utcDate, err := t.GetUTCDate()
 		if err != nil {
-			log.Errorf("Unable to get date from scope: %s", err.Error())
-		} else {
-			h, m, s := utcDate.Date()
-			isDST := '0'
-			if utcDate.IsDST() {
-				isDST = '1'
-			}
-
-			y, M, d := utcDate.Clock()
-			y = -2000 // need to output a single byte for the year
-			ret = fmt.Sprintf("%c%c%c%c%c%c%c%c#",
-				h, m, s, // H:M:S
-				M, d, y, // M:D:Y
-				0, // always UTC
-				isDST)
+			// not all mounts support this, so use our computer time I guess?
+			utcDate = time.Now()
 		}
+		h, m, s := utcDate.Date()
+		isDST := '0'
+		if utcDate.IsDST() {
+			isDST = '1'
+		}
+
+		y, M, d := utcDate.Clock()
+		y = -2000 // need to output a single byte for the year
+		ret = fmt.Sprintf("%c%c%c%c%c%c%c%c#",
+			h, m, s, // H:M:S
+			M, d, y, // M:D:Y
+			0, // always UTC
+			isDST)
 
 	case 'H':
 		// set date/time

--- a/telescope/nexstar.go
+++ b/telescope/nexstar.go
@@ -194,7 +194,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 			failed = true
 		}
 
-		long, err := t.GetSiteLatitude()
+		long, err := t.GetSiteLongitude()
 		if err != nil {
 			// logged at the end
 			failed = true
@@ -220,8 +220,8 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		// get date/time
 		utcDate, err := t.GetUTCDate()
 		if err != nil {
-			// not all mounts support this, so use our computer time I guess?
-			utcDate = time.Now()
+			log.Errorf("computer returned no UTC date: %s", err.Error())
+			return ""
 		}
 		h, m, s := utcDate.Date()
 		isDST := '0'

--- a/telescope/nexstar.go
+++ b/telescope/nexstar.go
@@ -65,12 +65,12 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 	switch buf[0] {
 	case 'K':
 		// echo next byte
-		ret = fmt.Sprintf("%c#", buf[1])
+		ret_val = []byte{buf[1], '#'}
 
 	case 'e', 'E':
 		ra, dec, err := t.GetRaDec()
 		if err != nil {
-			log.Errorf("Unable to get RA/DEC: %s", err.Error())
+			log.Errorf("unable to get RA/DEC: %s", err.Error())
 		} else {
 			radec := Coordinates{
 				RA:  ra,
@@ -88,7 +88,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		// Get AZM/ALT.  Note that AZM is 0->360, while Alt is -90->90
 		azm, alt, err := t.GetAzmAlt()
 		if err != nil {
-			log.Errorf("Unable to get AZM/ALT: %s", err.Error())
+			log.Errorf("unable to get AZM/ALT: %s", err.Error())
 		} else {
 			azm_int := uint32(azm / 360.0 * math.Pow(2, 16))
 			alt_int := uint32(alt / 360.0 * math.Pow(2, 16))
@@ -99,7 +99,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		// Get Precise AZM/ALT
 		azm, alt, err := t.GetAzmAlt()
 		if err != nil {
-			log.Errorf("Unable to get AZM/ALT: %s", err.Error())
+			log.Errorf("unable to get AZM/ALT: %s", err.Error())
 		} else {
 			azm_int := uint32(azm / 360.0 * 4294967296.0)
 			alt_int := uint32(alt / 360.0 * 4294967296.0)
@@ -110,9 +110,9 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		// get tracking mode
 		mode, err := t.GetTracking()
 		if err != nil {
-			log.Errorf("Unable to get tracking mode: %s", err.Error())
+			log.Errorf("unable to get tracking mode: %s", err.Error())
 		} else {
-			ret = fmt.Sprintf("%d#", mode)
+			ret = fmt.Sprintf("%c#", mode)
 		}
 
 	case 'T':
@@ -127,9 +127,36 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		ret = "50#"
 
 	case 'P':
-		// Slew
-		err = executeSlew(t, buf)
-		ret = "#"
+		// Pass through commands for Slew, GPS, RTC, etc
+		if int(buf[3]) == 254 {
+			// Get passthrough device version
+			switch int(buf[2]) {
+			case 16, 17:
+				// mounts
+				ret_val = []byte{5, 0, '#'}
+			case 176, 178:
+				// GPS & RTC
+				ret_val = []byte{1, 6, '#'}
+			default:
+				log.Errorf("invalid device version device type: %d", int(buf[3]))
+			}
+		} else {
+			switch int(buf[2]) {
+			case 176:
+				// GPS
+				ret_val, err = getGPS(t, buf)
+			case 178:
+				// RTC
+				ret_val, err = getRTC(t, buf)
+			case 16, 17:
+				err = executeSlew(t, buf)
+				ret = "#"
+			default:
+				log.Errorf("unsupported P command: %c%c%c%c%c%c%c",
+					buf[0], buf[1], buf[2], buf[3], buf[4], buf[5], buf[6])
+				ret = "#"
+			}
+		}
 
 	case 's':
 		// Precise sync aka: Align on object.  Uses the same math as 'e'
@@ -149,12 +176,12 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 			// auto-enable tracking?
 			mode, err := t.GetTracking()
 			if err != nil {
-				log.Errorf("Unable to get tracking mode: %s", err.Error())
+				log.Errorf("unable to get tracking mode: %s", err.Error())
 			} else {
 				if mode == alpaca.NotTracking {
 					err = t.PutTracking(alpaca.Alt_Az) // need any non-NotTracking value for true
 					if err != nil {
-						log.Errorf("Unable to auto-enable tracking: %s", err.Error())
+						log.Errorf("unable to auto-enable tracking: %s", err.Error())
 					}
 				}
 
@@ -170,12 +197,12 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 			// auto-enable tracking?
 			mode, err := t.GetTracking()
 			if err != nil {
-				log.Errorf("Unable to get tracking mode: %s", err.Error())
+				log.Errorf("unable to get tracking mode: %s", err.Error())
 			} else {
 				if mode == alpaca.NotTracking {
 					err = t.PutTracking(alpaca.Alt_Az) // need any non-NotTracking value for true
 					if err != nil {
-						log.Errorf("Unable to auto-enable tracking: %s", err.Error())
+						log.Errorf("unable to auto-enable tracking: %s", err.Error())
 					}
 				}
 
@@ -190,7 +217,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		failed := false
 		lat, err := t.GetSiteLatitude()
 		if err != nil {
-			log.Errorf("Error talking to scope: %s", err.Error())
+			log.Errorf("error talking to scope: %s", err.Error())
 			failed = true
 		}
 
@@ -210,7 +237,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		lat, long := NexstarToLatLong(buf[1:9])
 		err = t.PutSiteLatitude(lat)
 		if err != nil {
-			log.Errorf("Error talking to scope: %s", err.Error())
+			log.Errorf("talking to scope: %s", err.Error())
 		}
 		err = t.PutSiteLongitude(long)
 		// logged at the end
@@ -221,21 +248,21 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		utcDate, err := t.GetUTCDate()
 		if err != nil {
 			log.Errorf("computer returned no UTC date: %s", err.Error())
-			return ""
+			break
 		}
-		h, m, s := utcDate.Date()
-		isDST := '0'
+		h, m, s := utcDate.Clock()
+		var isDST byte = 0
 		if utcDate.IsDST() {
-			isDST = '1'
+			isDST = 1
 		}
 
-		y, M, d := utcDate.Clock()
-		y = -2000 // need to output a single byte for the year
-		ret = fmt.Sprintf("%c%c%c%c%c%c%c%c#",
-			h, m, s, // H:M:S
-			M, d, y, // M:D:Y
+		y, M, d := utcDate.Date()
+		y -= 2000 // need to output a single byte for the year
+		ret_val = []byte{
+			byte(h), byte(m), byte(s), // H:M:S
+			byte(M), byte(d), byte(y), // M:D:Y
 			0, // always UTC
-			isDST)
+			isDST, '#'}
 
 	case 'H':
 		// set date/time
@@ -254,13 +281,14 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 			int(buf[3]),        // sec S
 			0,                  // nanosec
 			tz)
+		log.Errorf("client set date to: %s", date.String())
 		err = t.PutUTCDate(date)
 		ret = "#"
 
 	case 'J':
 		// is alignment complete?
 		// since Alpaca has no similar command, aways return true
-		ret = "1#"
+		ret_val = []byte{1, '#'}
 
 	case 'L':
 		// Goto in progress??
@@ -272,18 +300,22 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 			ret = "0#"
 		}
 
+	case 'm':
+		// Model- hard code to 6/8 SE
+		ret_val = []byte{12, '#'}
+
 	case 'M':
 		// cancel GOTO
 		err = t.PutAbortSlew()
 		ret = "#"
 
 	default:
-		log.Errorf("Unsupported command: %c", buf[0])
+		log.Errorf("unsupported command: %c", buf[0])
 		ret = "#"
 	}
 
 	if err != nil {
-		log.Errorf("Error talking to scope: %s", err.Error())
+		log.Errorf("error talking to scope: %s", err.Error())
 	}
 
 	// convert our return string to the ret_val
@@ -319,7 +351,7 @@ func executeSlew(t *alpaca.Telescope, buf []byte) error {
 		// Alt/Dec
 		axis = alpaca.AxisAltDec
 	default:
-		log.Errorf("Unknown axis: %d", int(buf[2]))
+		log.Errorf("unknown axis: %d", int(buf[2]))
 	}
 
 	switch int(buf[3]) {
@@ -328,7 +360,7 @@ func executeSlew(t *alpaca.Telescope, buf []byte) error {
 	case 7, 37:
 		positive_direction = false
 	default:
-		log.Errorf("Unknown direction: %d", int(buf[3]))
+		log.Errorf("unknown direction: %d", int(buf[3]))
 	}
 
 	rate = nexstar_rate_to_ascom(positive_direction, int(buf[4]))
@@ -339,6 +371,97 @@ func executeSlew(t *alpaca.Telescope, buf []byte) error {
 
 	err := t.PutMoveAxis(axis, rate)
 	return err
+}
+
+/*
+ * getGPS is another 'P' command which returns date, location, etc from the GPS
+ * unit.  Note that the time & location values from these commands are supposed
+ * to come from the GPS and not RTC or hand controller, but ASCOM doesn't treat
+ * them differently since it is up to the driver.
+ *
+ * However, GPS was v1.6+ while hand controller was v2.3+ so we can expect
+ * some software to prefer/only support the GPS and not hand controller
+ * (Stellarium?)
+ */
+func getGPS(t *alpaca.Telescope, buf []byte) ([]byte, error) {
+	ret_val := []byte{}
+	switch int(buf[2]) {
+	case 55:
+		_, err := t.GetSiteLatitude()
+		if err != nil {
+			// GPS is not linked
+			return []byte{0, '#'}, nil
+		}
+		// GPS is linked
+		return []byte{1, '#'}, nil
+	case 1:
+		// Latitude
+		lat, err := t.GetSiteLatitude()
+		if err != nil {
+			log.Errorf("unable to GetSiteLatitude(): %s", err.Error())
+			return ret_val, err
+		}
+		ret_val = LatLongToGPS(lat)
+	case 2:
+		// Longitude
+		long, err := t.GetSiteLongitude()
+		if err != nil {
+			log.Errorf("unable to GetSiteLongitude(): %s", err.Error())
+			return ret_val, err
+		}
+		ret_val = LatLongToGPS(long)
+	case 3:
+		// Date: m, d
+		utcDate, err := t.GetUTCDate()
+		if err != nil {
+			log.Errorf("GPS returned no UTC date: %s", err.Error())
+			return ret_val, err
+		}
+		_, m, d := utcDate.Date()
+		ret_val = []byte{byte(m), byte(d), '#'}
+	case 4:
+		// Year: (x * 256) + y = year
+		utcDate, err := t.GetUTCDate()
+		if err != nil {
+			log.Errorf("GPS returned no UTC date: %s", err.Error())
+			return ret_val, err
+		}
+		year, _, _ := utcDate.Date()
+		var x int = year / 256
+		var y int = year % 256
+		ret_val = []byte{byte(x), byte(y), '#'}
+	case 51:
+		// Time: h, m, s
+		utcDate, err := t.GetUTCDate()
+		if err != nil {
+			log.Errorf("GPS returned no UTC date: %s", err.Error())
+			return ret_val, err
+		}
+		h, m, s := utcDate.Date()
+		ret_val = []byte{byte(h), byte(m), byte(s), '#'}
+	}
+	return ret_val, nil
+}
+
+/*
+ * getRTC is another 'P' command which returns date from the real time clock
+ * in the mount.  Note that the time values from these commands are supposed
+ * to come from the RTC and not GPS or hand controller, but ASCOM doesn't treat
+ * them differently since it is up to the driver.
+ *
+ * RTC is v1.6+ for get and v3.01+ for set.
+ */
+func getRTC(t *alpaca.Telescope, buf []byte) ([]byte, error) {
+	switch int(buf[2]) {
+	case 3, 4, 51:
+		// These commands to get date, time and year are the same
+		// as the GPS commands, so reuse that code
+		return getGPS(t, buf)
+	default:
+		log.Errorf("unsupported RTC P command: %c%c%c%c%c%c%c",
+			buf[0], buf[1], buf[2], buf[3], buf[4], buf[5], buf[6])
+	}
+	return []byte{}, nil
 }
 
 // Converts the direction & rate to an ASCOM rate
@@ -373,7 +496,7 @@ func NexstarToLatLong(b []byte) (float64, float64) {
 	return lat, long
 }
 
-// Convert Lat/Long to ABCDEFGH bytes
+// Convert Lat/Long to ABCDEFGH bytes for hand controller
 func LatLongToNexstar(lat float64, long float64) []byte {
 	var a, b, c, d, e, f, g, h byte
 	// West & South are negative
@@ -401,6 +524,28 @@ func LatLongToNexstar(lat float64, long float64) []byte {
 	g = byte(int(frac_minute * 60.0 * 60.0))
 
 	return []byte{a, b, c, d, e, f, g, h}
+}
+
+/*
+ * Convert a Lat OR Long to GPS fraction of a rotation "XYZ#"
+ *
+ * This is really a 24bit integer representing degrees as:
+ * (x*65536)+(y*256)+z / 2^24 * 360
+ */
+func LatLongToGPS(latlong float64) []byte {
+	var pos = make([]byte, 4)
+
+	// West & South are negative, so need to convert to positive degrees
+	if latlong < 0 {
+		latlong += 360.0
+	}
+
+	position := uint32(latlong * float64(2^24) / 360.0)
+	pos[0] = byte(position & 0x00ff0000 >> 16)
+	pos[1] = byte(position & 0x0000ff00 >> 8)
+	pos[2] = byte(position & 0x000000ff)
+	pos[3] = '#'
+	return pos
 }
 
 /*

--- a/telescope/nexstar.go
+++ b/telescope/nexstar.go
@@ -31,16 +31,22 @@ func (n *NexStar) HandleConnection(conn net.Conn, t *alpaca.Telescope) {
 			break
 		}
 		reply := n.nexstar_command(t, rlen, buf)
-		if len(reply) > 0 {
-			_, err = conn.Write(reply)
+		wlen := len(reply)
+		log.Debugf("our reply %d bytes: %v", wlen, reply)
+
+		if wlen > 0 {
+			x, err := conn.Write(reply)
 			if err != nil {
 				log.Errorf("writing reply to NexStar client: %s", err.Error())
+			} else if x != wlen {
+				log.Errorf("only wrote %d of %d bytes", x, wlen)
 			}
 		} else {
 			log.Errorf("command '%s' returned a zero length reply", string(buf))
 		}
-		rlen, err = conn.Read(buf)
+		rlen, err = conn.Read(buf) // blocks for next command
 	}
+
 	// Will get this any time the client sends a Fin, so don't log that
 	if err.Error() != "EOF" {
 		log.Errorf("conn.Read() returned error: %s", err.Error())
@@ -56,9 +62,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		for i := 1; i < len; i++ {
 			strbuf = fmt.Sprintf("%s %d", strbuf, buf[i])
 		}
-		if buf[0] != 'e' {
-			log.Debugf("Received %d bytes [%s]: %c %s", len, string(buf[:len]), buf[0], strbuf)
-		}
+		log.Debugf("Received %d bytes [%s]: %c %s", len, string(buf[:len]), buf[0], strbuf)
 	}
 
 	// single byte commands
@@ -217,7 +221,7 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		failed := false
 		lat, err := t.GetSiteLatitude()
 		if err != nil {
-			log.Errorf("error talking to scope: %s", err.Error())
+			log.Errorf("talking to scope: %s", err.Error())
 			failed = true
 		}
 
@@ -248,21 +252,29 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		utcDate, err := t.GetUTCDate()
 		if err != nil {
 			log.Errorf("computer returned no UTC date: %s", err.Error())
-			break
-		}
-		h, m, s := utcDate.Clock()
-		var isDST byte = 0
-		if utcDate.IsDST() {
-			isDST = 1
-		}
+			/*
+				ret_val = []byte{
+					0, 0, 0,
+					0, 0, 0,
+					0, 0, '#',
+				}
+			*/
+		} else {
+			h, m, s := utcDate.Clock()
+			var isDST byte = 0
+			if utcDate.IsDST() {
+				isDST = 1
+			}
 
-		y, M, d := utcDate.Date()
-		y -= 2000 // need to output a single byte for the year
-		ret_val = []byte{
-			byte(h), byte(m), byte(s), // H:M:S
-			byte(M), byte(d), byte(y), // M:D:Y
-			0, // always UTC
-			isDST, '#'}
+			y, M, d := utcDate.Date()
+			y -= 2000 // need to output a single byte for the year
+			ret_val = []byte{
+				byte(h), byte(m), byte(s), // H:M:S
+				byte(M), byte(d), byte(y), // M:D:Y
+				0, // always UTC
+				isDST, '#',
+			}
+		}
 
 	case 'H':
 		// set date/time

--- a/telescope/nexstar.go
+++ b/telescope/nexstar.go
@@ -216,6 +216,27 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 		// logged at the end
 		ret = "#"
 
+	case 'h':
+		// get date/time
+		utcDate, err := t.GetUTCDate()
+		if err != nil {
+			log.Errorf("Unable to get date from scope: %s", err.Error())
+		} else {
+			h, m, s := utcDate.Date()
+			isDST := '0'
+			if utcDate.IsDST() {
+				isDST = '1'
+			}
+
+			y, M, d := utcDate.Clock()
+			y = -2000 // need to output a single byte for the year
+			ret = fmt.Sprintf("%c%c%c%c%c%c%c%c#",
+				h, m, s, // H:M:S
+				M, d, y, // M:D:Y
+				0, // always UTC
+				isDST)
+		}
+
 	case 'H':
 		// set date/time
 		tz_val := int(buf[7])
@@ -235,6 +256,11 @@ func (n *NexStar) nexstar_command(t *alpaca.Telescope, len int, buf []byte) []by
 			tz)
 		err = t.PutUTCDate(date)
 		ret = "#"
+
+	case 'J':
+		// is alignment complete?
+		// since Alpaca has no similar command, aways return true
+		ret = "1#"
 
 	case 'L':
 		// Goto in progress??

